C error when assigning fixed array alias to struct optional field

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5ec9bb5.8835b6f
OS: linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P

getwd: /home/ge/Code/v/netaddr/issues
vexe: /home/ge/.vlang/v
vexe mtime: 2024-12-17 15:55:29

vroot: OK, value: /home/ge/.vlang
VMODULES: OK, value: /home/ge/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.47.1
Git vroot status: weekly.2024.51-9-g8835b6f2
.git/config present: true

CC version: cc (GCC) 14.2.1 20240910
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg struct_optional_aliased_fixed_array_field.vv`

```v
type Arr = [4]u8

struct Foo {
	bar int
	baz ?Arr
}

fn main() {
	f := Foo{
		bar: 1
		baz: Arr([u8(5), 4, 3, 2]!)
	}
	println(f)
	assert f.baz as Arr == [u8(5), 4, 3, 2]!
}
```

**What did you expect to see?**

Assertion passed, no C errors.

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/struct_optional_aliased_fixed_array_field.01JFAPX70SQ3AVKPSXNNJCDG8G.tmp.c:5919: error: invalid operand types for binary operation
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzYxYTYyMWYxNDRmNTg1YWU0ZGNjMmUiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.mvPYC8_YZwO_0-uS-NBKyOtns7f0MFjAItFrU-PYPFE">Huly&reg;: <b>V_0.6-21632</b></a></sub>

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -157,6 +157,18 @@ fn (mut g Gen) expr_with_opt(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type)
 	return ''
 }
 
+fn (mut g Gen) expr_with_fixed_array_opt(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type) {
+	tmp_var := g.new_tmp_var()
+	stmt_str := g.go_before_last_stmt().trim_space()
+	styp := g.styp(expr_typ)
+	g.empty_line = true
+	g.writeln('${styp} ${tmp_var} = {.state = 0};')
+	g.write('memcpy(${tmp_var}.data, ')
+	g.expr(expr)
+	g.writeln(', sizeof(${g.styp(ret_typ)}));')
+	g.write2(stmt_str, tmp_var)
+}
+
 fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 	mut node := unsafe { node_ }
 	if node.is_static {

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -688,13 +688,26 @@ fn (mut g Gen) struct_init_field(sfield ast.StructInitField, language ast.Langua
 		if field_unwrap_sym.info is ast.ArrayFixed {
 			match sfield.expr {
 				ast.Ident, ast.SelectorExpr {
-					g.fixed_array_var_init(g.expr_string(sfield.expr), is_auto_deref_var,
-						field_unwrap_sym.info.elem_type, field_unwrap_sym.info.size)
+					if sfield.expected_type.has_flag(.option) {
+						if field_unwrap_typ.has_flag(.option) {
+							g.expr_with_opt(sfield.expr, sfield.expected_type, field_unwrap_typ)
+						} else {
+							g.expr_with_fixed_array_opt(sfield.expr, sfield.expected_type,
+								field_unwrap_typ)
+						}
+					} else {
+						g.fixed_array_var_init(g.expr_string(sfield.expr), is_auto_deref_var,
+							field_unwrap_sym.info.elem_type, field_unwrap_sym.info.size)
+					}
 				}
 				ast.CastExpr, ast.CallExpr {
-					if sfield.expected_type.has_flag(.option) && !field_unwrap_typ.has_flag(.option) {
-						g.expr_with_fixed_array_opt(sfield.expr, sfield.expected_type,
-							field_unwrap_typ)
+					if sfield.expected_type.has_flag(.option) {
+						if field_unwrap_typ.has_flag(.option) {
+							g.expr_with_opt(sfield.expr, sfield.expected_type, field_unwrap_typ)
+						} else {
+							g.expr_with_fixed_array_opt(sfield.expr, sfield.expected_type,
+								field_unwrap_typ)
+						}
 					} else {
 						tmp_var := g.expr_with_var(sfield.expr, sfield.expected_type,
 							false)

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -692,9 +692,15 @@ fn (mut g Gen) struct_init_field(sfield ast.StructInitField, language ast.Langua
 						field_unwrap_sym.info.elem_type, field_unwrap_sym.info.size)
 				}
 				ast.CastExpr, ast.CallExpr {
-					tmp_var := g.expr_with_var(sfield.expr, sfield.expected_type, false)
-					g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
-						field_unwrap_sym.info.size)
+					if sfield.expected_type.has_flag(.option) && !field_unwrap_typ.has_flag(.option) {
+						g.expr_with_fixed_array_opt(sfield.expr, sfield.expected_type,
+							field_unwrap_typ)
+					} else {
+						tmp_var := g.expr_with_var(sfield.expr, sfield.expected_type,
+							false)
+						g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
+							field_unwrap_sym.info.size)
+					}
 				}
 				ast.ArrayInit {
 					if sfield.expr.has_index {

--- a/vlib/v/tests/structs/struct_field_init_with_fixed_array_opt_test.v
+++ b/vlib/v/tests/structs/struct_field_init_with_fixed_array_opt_test.v
@@ -6,10 +6,27 @@ struct Foo {
 }
 
 fn test_struct_field_init_with_fixed_array_opt() {
-	f := Foo{
+	f1 := Foo{
 		bar: 1
 		baz: Arr([u8(5), 4, 3, 2]!)
 	}
-	println(f)
-	assert f.baz as Arr == [u8(5), 4, 3, 2]!
+	println(f1)
+	assert f1.baz as Arr == [u8(5), 4, 3, 2]!
+
+	f2 := Foo{
+		bar: 1
+		baz: ?Arr(none)
+	}
+	println(f2)
+	assert f2.bar == 1
+	assert f2.baz == none
+
+	arr := Arr([u8(5), 4, 3, 2]!)
+	f3 := Foo{
+		bar: 1
+		baz: arr
+	}
+	println(f3)
+	assert f3.bar == 1
+	assert f3.baz as Arr == [u8(5), 4, 3, 2]!
 }

--- a/vlib/v/tests/structs/struct_field_init_with_fixed_array_opt_test.v
+++ b/vlib/v/tests/structs/struct_field_init_with_fixed_array_opt_test.v
@@ -1,0 +1,15 @@
+type Arr = [4]u8
+
+struct Foo {
+	bar int
+	baz ?Arr
+}
+
+fn test_struct_field_init_with_fixed_array_opt() {
+	f := Foo{
+		bar: 1
+		baz: Arr([u8(5), 4, 3, 2]!)
+	}
+	println(f)
+	assert f.baz as Arr == [u8(5), 4, 3, 2]!
+}

